hazelcast/src/main/java/com/hazelcast/jet/impl/ClientJobProxy.java

```diff
@@ -256,12 +256,16 @@ protected boolean isRunning() {
 
     @Nonnull
     @Override
-    public UUID addStatusListener(@Nonnull JobStatusListener listener) {
+    protected UUID doAddStatusListener(JobStatusListener listener) {
         requireNonNull(listener, "Listener cannot be null");
-        ClientJobStatusEventHandler handler = new ClientJobStatusEventHandler(listener);
-        handler.registrationId = container().getListenerService().registerListener(
-                createJobStatusListenerCodec(getId()), handler);
-        return handler.registrationId;
+        try {
+            ClientJobStatusEventHandler handler = new ClientJobStatusEventHandler(listener);
+            handler.registrationId = ((ClientListenerServiceImpl) container().getListenerService())
+                    .registerListener(createJobStatusListenerCodec(getId()), handler, coordinatorId());
+            return handler.registrationId;
+        } catch (Throwable t) {
+            throw rethrow(t.getCause());
```

Comment:
`ClientListenerServiceImpl` wraps the member-side exception with an `HazelcastException("Listener can not be added", e)`.

Reply:
it is not defined in contract for `ClientListenerService.registerListener`, so it not safe to rely on that

Reply:
this is unexpected error so we are allowed to throw `HazelcastException`

Reply:
`JobCoordinationService` might throw a `JobNotFoundException` if the request-initiating job proxy is outdated. This error is translated to `IllegalStateException` by the proxy, so it is not unexpected. `ClientListenerServiceImpl`'s wrapping is unnecessary. Maybe we can peel the exception conditionally to make it more robust, but first the existing behavior should be justified.

Reply:
what exactly will be thrown by `doAddStatusListener`  in case you described without `rethrow(t.getCause())`? Will it be: `HazelcastException("Listener can not be added")` caused by `JobNotFoundException`?

> Maybe we can peel the exception conditionally to make it more robust

let's do this. Maybe it would better to do such exception unwrapping in `addStatusListener`?
BTW, `doAddStatusListener` javadoc should not define what `addStatusListener` does with exceptions.